transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -392,10 +392,20 @@ def add_trace(self, trace):
         # FIXME, add an exponential polling backoff so that we wait 1, 2, 4, 8, etc ms
         self._ensure_connection(conn_id, True)
 
-        response = await self._jlink_async.send_rpc(
-            self._device_info, self._control_info,
-            address, rpc_id, payload, timeout)
-        return response['payload']
+        response = None
+        
+        for attempt in range(retries):
+            try:
+                response = await self._jlink_async.send_rpc(
+                    self._device_info, self._control_info,
+                    address, rpc_id, payload, timeout)
+            except:
```

Comment:
This should only catch `ExternalGate` errors since that's all we are retrying, it shouldn't chomp all RPC errors (such as timeouts or disconnections).  You may need to modify where the exception is thrown to distinguish it with a different internal class.

You should log with `debug` level as well when you retry since its expected to happen.

Finally if you get more than max retries attempts and don't get the gate, you should raise a `DeviceAdapterError` indicating failure, not just chomp the error.

Reply:
Understood. I pushed a fix that reflects the behavior you just mentioned. So if it receives a `HardwareError` it will check if its an external gate error. If its not, it'll re-raise the exception higher. If it catches any other exception, it will just re-raise as well. Finally, when it finishes the loop after all the retries, it will raise a `DeviceAdapterError`.